### PR TITLE
[GLUTEN-1654] [VL] support approx_count_distinct for velox

### PR DESCRIPTION
**Pull request by zhli1142015**

### CODE REVIEW

**Comment by zhouyuan** [comment]
Run Gluten Clickhouse CI

**Comment by zhli1142015** [comment]
> @zhli1142015 there's a minor commit in your repo here, should also PR to oap/velox? https://github.com/zhli1142015/velox/tree/whitelist-approx_distinct
> 
> -yuan

sure, thanks, here is the PR: https://github.com/oap-project/velox/pull/270

File: gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala [hunk 1]
```diff
@@ -593,6 +593,21 @@ abstract class HashAggregateExecBaseTransformer(
             case other =>
               throw new UnsupportedOperationException(s"not currently supported: $other.")
           }
+        case hllAdapter: HLLVeloxAdapter =>
```
Comment (jinchengchenghh) [inline gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala:596]: gluten-core is a common module for all the backends, can we put this code to velox backend modules?

Comment (zhli1142015) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala]: Updated, thanks.

File: gluten-core/src/main/scala/io/glutenproject/expression/ExpressionMappings.scala [hunk 2]
```diff
@@ -427,7 +428,8 @@ object ExpressionMappings {
     Sig[CovPopulation](COVAR_POP),
     Sig[CovSample](COVAR_SAMP),
     Sig[Last](LAST),
-    Sig[First](FIRST)
+    Sig[First](FIRST),
+    Sig[HLLVeloxAdapter](APPROX_DISTINCT)
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/expression/ExpressionMappings.scala:432]: Add special function signature for each backend is supported. Do we consider moving this to Velox backend only?

File: gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala (hunk 1 above)
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala]: +1
Still observed several similar issues.

File: gluten-core/src/main/scala/io/glutenproject/expression/ExpressionMappings.scala (hunk 2 above)
Comment (zhli1142015) [reply gluten-core/src/main/scala/io/glutenproject/expression/ExpressionMappings.scala]: Moved to velox, please help check if this is right approach.

File: gluten-core/src/main/scala/org/apache/spark/sql/catalyst/VeloxAggregateFunctionRewriteRule.scala [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions.Literal
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, HLLVeloxAdapter, HyperLogLogPlusPlus}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types._
+import io.glutenproject.backendsapi.BackendsApiManager
+import io.glutenproject.GlutenConfig
+
+case class VeloxAggregateFunctionRewriteRule(spark: SparkSession) extends Rule[LogicalPlan] {
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply gluten-core/src/main/scala/org/apache/spark/sql/catalyst/VeloxAggregateFunctionRewriteRule.scala]: sure, moved.

File: gluten-core/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HLLVeloxAdapter.scala [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.trees.BinaryLike
+import org.apache.spark.sql.catalyst.util.{ArrayData, HyperLogLogPlusPlusHelper}
+import org.apache.spark.sql.types._
+
+// HLL in velox's intermediate type is binary, which is different from spark HLL.
+// We add this wrapper to align the intermediate type for HLL functions.
+case class HLLVeloxAdapter(
```
[the remark this answers is not in the document]
Comment (zhli1142015) [reply gluten-core/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HLLVeloxAdapter.scala]: moved, thanks

File: gluten-core/src/main/scala/io/glutenproject/expression/ExpressionMappings.scala (hunk 2 above)
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ExpressionMappings.scala]: Thank you. LGTM.

**Review by rui-mo — APPROVED** [review]

File: backends-velox/src/main/scala/org/apache/spark/sql/catalyst/VeloxAggregateFunctionRewriteRule.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import io.glutenproject.GlutenConfig
+
+case class VeloxAggregateFunctionRewriteRule(spark: SparkSession) extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
+        case a: Aggregate =>
+          a.transformExpressions {
+            case hllExpr @ AggregateExpression(hll: HyperLogLogPlusPlus, _, _, _, _)
+              if GlutenConfig.getConf.enableNativeHyperLogLogAggregateFunction &&
+                GlutenConfig.getConf.enableColumnarHashAgg &&
+                !hasDistinctAggregateFunc(a) && isDataTypeSupported(hll.child.dataType) =>
```
Comment (boneanxs) [inline backends-velox/src/main/scala/org/apache/spark/sql/catalyst/VeloxAggregateFunctionRewriteRule.scala:36]: Hey @zhli1142015 @zhouyuan @rui-mo , I must miss something here, curious why this function doesn't support with `distinct`, is that `PartialMerge` is not supported?

Comment (zhli1142015) [reply backends-velox/src/main/scala/org/apache/spark/sql/catalyst/VeloxAggregateFunctionRewriteRule.scala:36]: I remember PartialMerge is not well supported when doing this PR. this check is removed already in main branch.

Comment (boneanxs) [reply backends-velox/src/main/scala/org/apache/spark/sql/catalyst/VeloxAggregateFunctionRewriteRule.scala:36]: I see, thank you for point me out!

**Pull request merged**